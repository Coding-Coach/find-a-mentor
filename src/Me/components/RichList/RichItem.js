import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

const RichItem = ({
  id,
  avatar,
  title,
  subtitle,
  tag,
  info,
  expand,
  onClick,
  children,
}) => {
  return (
    <Root highlight={!!children && !expand} expanded={expand}>
      <Main onClick={() => onClick(id)}>
        <ItemRow>
          <ItemAvatar>{avatar && <img src={avatar} alt="avatar" />}</ItemAvatar>
          <Titles>
            <ItemRow>
              <Title>{title}</Title>
              <Tag theme={tag.theme}>{tag.value}</Tag>
            </ItemRow>
            <Subtitle>{subtitle}</Subtitle>
          </Titles>
          <Info>{info}</Info>
        </ItemRow>
      </Main>
      <Content expand={expand}>{children}</Content>
    </Root>
  );
};

RichItem.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  avatar: PropTypes.string,
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string,
  info: PropTypes.string,
  tag: PropTypes.shape({
    value: PropTypes.string.isRequired,
    theme: PropTypes.oneOf([
      'primary',
      'secondary',
      'danger',
      'checked',
      'disabled',
    ]),
  }),
  onClick: PropTypes.func,
  expand: PropTypes.bool,
};

const themeColours = {
  primary: '#69D5B1',
  secondary: '#F3CA3E',
  danger: '#FF5F58',
  checked: '#69d579',
  disabled: 'e0e0e0',
};

// Styled Components

const ItemRow = styled.div`
  display: flex;
  gap: 5px;
`;

const ItemCol = styled.div`
  display: flex;
  flex-direction: column;
`;

const Main = styled(ItemRow)`
  height: 80px;
  padding: 0 1px;
  align-items: center;
  > div {
    flex: 1;
  }
`;

const Titles = styled(ItemCol)`
  flex: 1;

  ${ItemRow} {
    align-items: center;
  }
`;

const ItemAvatar = styled.div`
  user-select: none;
  width: 45px;
  overflow: hidden;
  display: flex;
  align-items: center;

  img {
    width: 100%;
    border-radius: 50%;
  }
`;

const Title = styled.h5`
  width: 94px;
  margin: 0 10px 0 0;
  font-size: 14px;
  font-weight: normal;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const Subtitle = styled(Title)`
  color: #828282;
`;

const Info = styled.div`
  display: flex;
  justify-content: center;
`;

const Tag = styled.div`
  width: 46px;
  color: #fff;
  border-radius: 3px;
  padding: 1px 6px;
  user-select: none;
  font-size: 8px;
  line-height: 12px;
  text-align: center;
  background-color: ${({ theme }) => themeColours[theme]};
`;

const Content = styled.div`
  transition: max-height 350ms ease, padding 300ms 50ms, opacity 300ms;
  max-height: 0;
  opacity: 0;
  visibility: hidden;
  overflow-y: auto;

  ${({ expand }) => ({
    visibility: 'visible',
    paddingBottom: expand ? '30px' : 0,
    maxHeight: expand ? '500px' : 0,
    opacity: expand ? 1 : 0,
  })}
`;

const Root = styled.div`
  position: relative;
  margin: 0 calc(-1 * var(--padding-inline));
  padding: 0 var(--padding-inline);
  transition: background-color 0.15s ease;

  &:hover {
    background-color: ${({ highlight }) => (highlight ? '#f2f2f2' : '')};
    ${Main} {
      cursor: ${({ highlight, expanded }) =>
        expanded || highlight ? 'pointer' : 'default'};
    }
  }
  & + &::before {
    content: ' ';
    border-top: 1px solid #f2f2f2;
    display: block;
    position: absolute;
    width: 100%;
    top: -1px;
  }
`;

export default RichItem;
