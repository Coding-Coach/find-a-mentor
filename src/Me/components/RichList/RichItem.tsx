import { FC } from 'react';
import styled from 'styled-components';
import { RichItemProps, RichItemTagTheme } from './ReachItemTypes';

const RichItem: FC<RichItemProps> = ({
  id,
  avatar,
  title,
  subtitle,
  tag,
  info,
  expand,
  onClick,
  children,
}) => {
  return (
    <Root highlight={!!children && !expand} expanded={expand}>
      <Main onClick={() => onClick(id)}>
        <ItemRow>
          <ItemAvatar>{avatar && <img src={avatar} alt="avatar" />}</ItemAvatar>
          <Titles>
            <ItemRow>
              <Title>{title}</Title>
              <Tag role="status" theme={tag.theme}>
                {tag.value}
              </Tag>
            </ItemRow>
            <Subtitle>{subtitle}</Subtitle>
          </Titles>
          <Info>{info}</Info>
        </ItemRow>
      </Main>
      <Content expand={expand}>{children}</Content>
    </Root>
  );
};

const themeColours: { [key in RichItemTagTheme]: string } = {
  primary: '#69d5b1',
  secondary: '#f3ca3e',
  danger: '#ff5f58',
  checked: '#69d579',
  disabled: '#e0e0e0',
  cancelled: '#ff5f58',
};

const ItemRow = styled.div`
  display: flex;
  gap: 5px;
`;

const ItemCol = styled.div`
  display: flex;
  flex-direction: column;
`;

const Main = styled(ItemRow)`
  height: 80px;
  padding: 0 1px;
  align-items: center;
  > div {
    flex: 1;
  }
`;

const Titles = styled(ItemCol)`
  flex: 1;

  ${ItemRow} {
    align-items: center;
  }
`;

const ItemAvatar = styled.div`
  user-select: none;
  width: 45px;
  overflow: hidden;
  display: flex;
  align-items: center;

  img {
    width: 100%;
    border-radius: 50%;
  }
`;

const Title = styled.h5`
  width: 94px;
  margin: 0 10px 0 0;
  font-size: 14px;
  font-weight: normal;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const Subtitle = styled(Title)`
  color: #828282;
`;

const Info = styled.div`
  display: flex;
  justify-content: center;
`;

type TagTheme = keyof typeof themeColours;

const Tag = styled.div`
  width: 46px;
  color: #fff;
  border-radius: 3px;
  padding: 1px 6px;
  user-select: none;
  font-size: 8px;
  line-height: 12px;
  text-align: center;
  background-color: ${({ theme }: { theme: TagTheme }) => themeColours[theme]};
`;

const Content = styled.div<{ expand: boolean }>`
  transition: max-height 350ms ease, padding 300ms 50ms, opacity 300ms;
  max-height: 0;
  opacity: 0;
  visibility: hidden;
  overflow-y: auto;

  ${({ expand }) => ({
    visibility: 'visible',
    paddingBottom: expand ? '30px' : 0,
    maxHeight: expand ? '500px' : 0,
    opacity: expand ? 1 : 0,
  })}
`;

const Root = styled.div<{ highlight: boolean; expanded: boolean }>`
  position: relative;
  margin: 0 calc(-1 * var(--padding-inline));
  padding: 0 var(--padding-inline);
  transition: background-color 0.15s ease;

  &:hover {
    background-color: ${({ highlight }) => (highlight ? '#f2f2f2' : '')};
    ${Main} {
      cursor: ${({ highlight, expanded }) =>
        expanded || highlight ? 'pointer' : 'default'};
    }
  }
  & + &::before {
    content: ' ';
    border-top: 1px solid #f2f2f2;
    display: block;
    position: absolute;
    width: 100%;
    top: -1px;
  }
`;

export default RichItem;
