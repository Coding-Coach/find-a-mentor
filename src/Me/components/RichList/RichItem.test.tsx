import { render } from '@testing-library/react';
import RichItem, { RichItemProps } from './RichItem';

const defaults: RichItemProps = {
  id: 1,
  expand: true,
  avatar: '../../assets/me/mentors.svg',
  title: 'Mooli m',
  subtitle: 'web dev',
  tag: {
    value: 'pending',
    theme: 'secondary',
  },
  info: '3 days ago',
  onClick: jest.fn(),
};

const init = (
  overrides?: Partial<RichItemProps & { children: React.ReactNode }>
) => {
  const props: RichItemProps = {
    ...defaults,
    ...overrides,
  };

  const api = render(<RichItem {...props} />);

  return {
    props,
    ...api,
  };
};

describe('RichItem component', () => {
  it('Should render all props correctly', () => {
    const { getByText, getByAltText, props } = init();

    getByAltText('avatar');
    getByText(props.title);
    getByText(props.subtitle);
    getByText(props.tag.value);
    getByText(props.info);
  });
  it('renders primary tag when specified with theme prop', () => {
    const { container } = init({
      title: 'Accepted',
      tag: {
        value: 'Accepted',
        theme: 'primary',
      },
    });
    expect(container).toMatchSnapshot();
  });
  it('renders secondary tag when specified with theme prop', () => {
    const { container } = init({
      title: 'Pending',
      tag: {
        value: 'Pending',
        theme: 'secondary',
      },
    });
    expect(container).toMatchSnapshot();
  });

  it('renders danger tag when specified with theme prop', () => {
    const { container } = init({
      title: 'Declined',
      tag: {
        value: 'Declined',
        theme: 'danger',
      },
    });
    expect(container).toMatchSnapshot();
  });
});

describe('RichItem component with children', () => {
  it('should not show children content if "expand" is falsey', () => {
    const { getByTestId } = init({
      children: <div data-testid="content">Hi ðŸ˜Ž</div>,
      expand: false,
    });

    expect(getByTestId('content')).not.toBeVisible();
  });

  it('should show children content if "expand" is truthy', async () => {
    const { getByTestId } = init({
      children: <div data-testid="content">Hi ðŸ˜Ž</div>,
    });

    expect(getByTestId('content')).toBeVisible();

    // const myRender = (expand: boolean) => (
    //   <RichItem {...item} expand={expand}>
    //     <div data-testid="content">Hi ðŸ˜Ž</div>
    //   </RichItem>
    // );
    // const { getByTestId, queryByTestId, rerender } = render(myRender());

    // await waitFor(() => {
    //   expect(queryByTestId('content')).not.toBeVisible();
    // });

    // rerender(myRender(true));

    // await waitFor(() => {
    //   expect(getByTestId('content')).toBeVisible();
    // });
  });
});
