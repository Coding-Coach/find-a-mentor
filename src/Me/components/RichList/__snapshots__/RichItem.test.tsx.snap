// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichItem component renders danger tag when specified with theme prop 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 5px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  height: 80px;
  padding: 0 1px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 > div {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c8 .c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 45px;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 img {
  width: 100%;
  border-radius: 50%;
}

.c9 {
  width: 94px;
  margin: 0 10px 0 0;
  font-size: 14px;
  font-weight: normal;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c11 {
  color: #828282;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c10 {
  width: 46px;
  color: #fff;
  border-radius: 3px;
  padding: 1px 6px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 8px;
  line-height: 12px;
  text-align: center;
  background-color: #ff5f58;
}

.c13 {
  -webkit-transition: max-height 350ms ease,padding 300ms 50ms,opacity 300ms;
  transition: max-height 350ms ease,padding 300ms 50ms,opacity 300ms;
  max-height: 0;
  opacity: 0;
  visibility: hidden;
  overflow-y: auto;
  visibility: visible;
  padding-bottom: 30px;
  max-height: 500px;
  opacity: 1;
}

.c1 {
  position: relative;
  margin: 0 calc(-1 * var(--padding-inline));
  padding: 0 var(--padding-inline);
  -webkit-transition: background-color 0.15s ease;
  transition: background-color 0.15s ease;
}

.c1:hover .c3 {
  cursor: pointer;
}

.c0 + .c0::before {
  content: ' ';
  border-top: 1px solid #f2f2f2;
  display: block;
  position: absolute;
  width: 100%;
  top: -1px;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2 c3 c4 c5"
    >
      <div
        class="c2 c4"
      >
        <div
          class="c6"
        >
          <img
            alt="avatar"
            src="../../assets/me/mentors.svg"
          />
        </div>
        <div
          class="c7 c8"
        >
          <div
            class="c2 c4"
          >
            <h5
              class="c9"
            >
              Declined
            </h5>
            <div
              class="c10"
              role="status"
            >
              Declined
            </div>
          </div>
          <h5
            class="c9 c11"
          >
            web dev
          </h5>
        </div>
        <div
          class="c12"
        >
          3 days ago
        </div>
      </div>
    </div>
    <div
      class="c13"
    />
  </div>
</div>
`;

exports[`RichItem component renders primary tag when specified with theme prop 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 5px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  height: 80px;
  padding: 0 1px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 > div {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c8 .c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 45px;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 img {
  width: 100%;
  border-radius: 50%;
}

.c9 {
  width: 94px;
  margin: 0 10px 0 0;
  font-size: 14px;
  font-weight: normal;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c11 {
  color: #828282;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c10 {
  width: 46px;
  color: #fff;
  border-radius: 3px;
  padding: 1px 6px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 8px;
  line-height: 12px;
  text-align: center;
  background-color: #69d5b1;
}

.c13 {
  -webkit-transition: max-height 350ms ease,padding 300ms 50ms,opacity 300ms;
  transition: max-height 350ms ease,padding 300ms 50ms,opacity 300ms;
  max-height: 0;
  opacity: 0;
  visibility: hidden;
  overflow-y: auto;
  visibility: visible;
  padding-bottom: 30px;
  max-height: 500px;
  opacity: 1;
}

.c1 {
  position: relative;
  margin: 0 calc(-1 * var(--padding-inline));
  padding: 0 var(--padding-inline);
  -webkit-transition: background-color 0.15s ease;
  transition: background-color 0.15s ease;
}

.c1:hover .c3 {
  cursor: pointer;
}

.c0 + .c0::before {
  content: ' ';
  border-top: 1px solid #f2f2f2;
  display: block;
  position: absolute;
  width: 100%;
  top: -1px;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2 c3 c4 c5"
    >
      <div
        class="c2 c4"
      >
        <div
          class="c6"
        >
          <img
            alt="avatar"
            src="../../assets/me/mentors.svg"
          />
        </div>
        <div
          class="c7 c8"
        >
          <div
            class="c2 c4"
          >
            <h5
              class="c9"
            >
              Accepted
            </h5>
            <div
              class="c10"
              role="status"
            >
              Accepted
            </div>
          </div>
          <h5
            class="c9 c11"
          >
            web dev
          </h5>
        </div>
        <div
          class="c12"
        >
          3 days ago
        </div>
      </div>
    </div>
    <div
      class="c13"
    />
  </div>
</div>
`;

exports[`RichItem component renders secondary tag when specified with theme prop 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 5px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  height: 80px;
  padding: 0 1px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 > div {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c8 .c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 45px;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 img {
  width: 100%;
  border-radius: 50%;
}

.c9 {
  width: 94px;
  margin: 0 10px 0 0;
  font-size: 14px;
  font-weight: normal;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c11 {
  color: #828282;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c10 {
  width: 46px;
  color: #fff;
  border-radius: 3px;
  padding: 1px 6px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 8px;
  line-height: 12px;
  text-align: center;
  background-color: #f3ca3e;
}

.c13 {
  -webkit-transition: max-height 350ms ease,padding 300ms 50ms,opacity 300ms;
  transition: max-height 350ms ease,padding 300ms 50ms,opacity 300ms;
  max-height: 0;
  opacity: 0;
  visibility: hidden;
  overflow-y: auto;
  visibility: visible;
  padding-bottom: 30px;
  max-height: 500px;
  opacity: 1;
}

.c1 {
  position: relative;
  margin: 0 calc(-1 * var(--padding-inline));
  padding: 0 var(--padding-inline);
  -webkit-transition: background-color 0.15s ease;
  transition: background-color 0.15s ease;
}

.c1:hover .c3 {
  cursor: pointer;
}

.c0 + .c0::before {
  content: ' ';
  border-top: 1px solid #f2f2f2;
  display: block;
  position: absolute;
  width: 100%;
  top: -1px;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2 c3 c4 c5"
    >
      <div
        class="c2 c4"
      >
        <div
          class="c6"
        >
          <img
            alt="avatar"
            src="../../assets/me/mentors.svg"
          />
        </div>
        <div
          class="c7 c8"
        >
          <div
            class="c2 c4"
          >
            <h5
              class="c9"
            >
              Pending
            </h5>
            <div
              class="c10"
              role="status"
            >
              Pending
            </div>
          </div>
          <h5
            class="c9 c11"
          >
            web dev
          </h5>
        </div>
        <div
          class="c12"
        >
          3 days ago
        </div>
      </div>
    </div>
    <div
      class="c13"
    />
  </div>
</div>
`;
